examples/ct/ct_handlers.go

```diff
@@ -202,3 +340,68 @@ func (c CTRequestHandlers) RegisterCTHandlers() {
 	http.HandleFunc(pathFor("get-roots"), wrappedGetRootsHandler(c.trustedRoots, c.rpcClient))
 	http.HandleFunc(pathFor("get-entry-and-proof"), wrappedGetEntryAndProofHandler(c.rpcClient))
 }
+
+// Generates a custom error page to give more information on why something didn't work
+// TODO(Martin2112): Not sure if we want to expose any detail or not
+func sendHttpError(w http.ResponseWriter, statusCode int, err error) {
+	http.Error(w, fmt.Sprintf("%s\n%v", http.StatusText(statusCode), err), statusCode)
+}
+
+// getRPCDeadlineTime calculates the future time an RPC should expire based on our config
+func getRPCDeadlineTime(c CTRequestHandlers) time.Time {
+	return c.timeSource.Now().Add(c.rpcDeadline)
+}
+
+func rpcStatusOK(status *trillian.TrillianApiStatus) bool {
+	return status != nil && status.StatusCode == trillian.TrillianApiStatusCode_OK
+}
+
+// verifyAddChain is used by add-chain and add-pre-chain. It does the checks that the supplied
+// cert is of the correct type and chains to a trusted root.
+// TODO(Martin2112): This may not implement all the RFC requirements. Check what is provided
+// by fixchain (called by this code) plus the ones here to make sure that it is compliant.
+func verifyAddChain(req addChainRequest, w http.ResponseWriter, trustedRoots PEMCertPool, expectingPrecert bool) []*x509.Certificate {
+	// We already checked that the chain is not empty so can move on to verification
+	validPath, err := ValidateChain(req.Chain, trustedRoots)
+
+	if err != nil {
+		// We rejected it because the cert failed checks or we could not find a path to a root etc.
+		// Lots of possible causes for errors
+		glog.Warningf("Chain failed to verify: %v", req)
+		sendHttpError(w, http.StatusBadRequest, err)
+		return nil
+	}
+
+	isPrecert, err := IsPrecertificate(validPath[0])
+
+	// The type of the leaf must match the one the handler expects
+	if isPrecert != expectingPrecert || err != nil {
+		if expectingPrecert {
+			glog.Warningf("Cert (or precert with invalid CT ext) submitted as precert chain: %v", req)
+		} else {
+			glog.Warningf("Precert (or cert with invalid CT ext) submitted as cert chain: %v", req)
+		}
+		sendHttpError(w, http.StatusBadRequest, err)
+		return nil
+	}
+
+	return validPath
+}
+
+// marshalLogIDAndSignatureForResponse is used by add-chain and add-pre-chain. It formats the
+// signature and log id ready to send to the client.
+func marshalLogIDAndSignatureForResponse(sct ct.SignedCertificateTimestamp, km crypto.KeyManager) ([sha256.Size]byte, string, error) {
```

Comment:
It might be better to pass `sct` and `km` as pointers, to avoid copying, though this probably has ramifications on other function signatures so feel free to defer/ignore this.